programs/native/template_signer/src/lib.rs

```diff
@@ -0,0 +1,579 @@
+    /// SigningContract
+    InitEscrow(u64),
+
+    /// * account[0] - must be signed, current owner
+    /// * account[1] - the SigningContract
+    /// * account[2] - the new owner
+    ///
+    /// This should be executed in an atomic transaction with
+    /// Instruction 0, System::Move tokens from new owner to current owner
+    /// Instruction 1, Signer::ChangeOwner
+    ChangeOwner,
+
+    /// * account[0] - must be signed, current owner
+    /// * account[1] - the SigningContract
+    ///
+    /// Message respond with a signed tx.  If the tx is valid, the state is moved to Signed
+    Request(Request),
+
+    /// * account[0] - any caller
+    /// * account[1] - the SigningContract
+    ///
+    /// Valid signature that is a response
+    SignatureResponse(Vec<u8>),
+
+    /// * account[0] - any caller
+    /// * account[1] - the SigningContract
+    ///
+    /// Signature prooves that there was a double spend
+    ConcurrentSignature { msg: Vec<u8>, sig: Vec<u8> },
+
+    /// * account[0] - caller
+    /// * account[1] - the SigningContract
+    ///
+    /// Message to claim the tokens
+    /// * (State::Created | State::EscrowFilled | State::Claimed) & `template_timeout`
+    ///    The `signer` can claim `guarantee`, the `owner` can claim the `escrow`.    
+    /// * Requested & `requested_timeout`
+    ///    The the `owner` can claim the `escrow` and the `guarantee`.    
+    /// * Signed & `signed_timeout`
+    ///    The the `signer` can claim the `escrow` and the `guarantee`.    
+    /// * ConcurrentSignature
+    ///    The the `owner` can claim the `escrow` and the `guarantee`.    
+    /// Moves the state to State::Claimed
+    Claim,
+}
+
+impl Key {
+    fn check_sig(&self, _msg: &[u8], _sig: &[u8]) -> bool {
+        //TODO: implement this
+        true
+    }
+}
+impl Template {
+    fn verify_request(&self, request: &Request) -> Result<(), Error> {
+        if request.msg.len() != self.msg.len() {
+            return Err(Error::Error);
+        }
+        if request.msg.len() != self.mask.len() {
+            return Err(Error::Error);
+        }
+        for ((r, m), mask) in request.msg.iter().zip(&self.msg).zip(&self.mask) {
+            if *r & *mask != *m & *mask {
+                return Err(Error::Error);
+            }
+        }
+        Ok(())
+    }
+}
+impl SigningContract {
+    fn template_timeout(&self) -> bool {
+        //TODO: transaction height is necessary for this check
+        false
+    }
+    fn requested_timeout(&self) -> bool {
+        //TODO: transaction height is necessary for this check
+        false
+    }
+    fn signed_timeout(&self) -> bool {
+        //TODO: transaction height is necessary for this check
+        false
+    }
+    fn init(&self, keyed_accounts: &mut Vec<KeyedAccount>) -> Result<(), Error> {
+        if !(self.escrow == 0 && self.state == State::Created) {
+            return Err(Error::Error);
+        }
+        //TODO: keyed_accounts[0].is_signed()?;
```

Comment:
These TODOs can be fixed now, SDK now surfaces the keys that signed the transaction